cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationRunDetail.java

```diff
@@ -92,100 +99,108 @@ public boolean equals(Object o) {
+        getPullAppsRequest());
   }
 
   /**
    * Builder to create a OperationRunDetail.
    */
-  public static <T> Builder<T> builder() {
-    return new Builder<>();
+  public static Builder builder() {
+    return new Builder();
   }
 
   /**
    * Builder to create OperationRunDetail.
    *
    * @param detail existing detail to copy fields from.
    */
-  public static <T> Builder<T> builder(OperationRunDetail<T> detail) {
-    return new Builder<>(detail);
+  public static Builder builder(OperationRunDetail detail) {
+    return new Builder(detail);
   }
 
   /**
    * Builds RunRecordMetas.
    */
-  public static class Builder<T> {
+  public static class Builder {
 
     protected OperationRunId runId;
     protected OperationRun run;
     protected byte[] sourceId;
     protected String principal;
-    protected T request;
+    protected PullAppsRequest pullAppsRequest;
 
     protected Builder() {
     }
 
-    protected Builder(OperationRunDetail<T> detail) {
+    protected Builder(OperationRunDetail detail) {
       sourceId = detail.getSourceId();
       principal = detail.getPrincipal();
-      request = detail.getRequest();
       run = detail.getRun();
       runId = detail.getRunId();
+      pullAppsRequest = detail.getPullAppsRequest();
     }
 
-    public Builder<T> setSourceId(byte[] sourceId) {
+    public Builder setSourceId(byte[] sourceId) {
       this.sourceId = sourceId;
       return this;
     }
 
-    public Builder<T> setPrincipal(String principal) {
+    public Builder setPrincipal(String principal) {
       this.principal = principal;
       return this;
     }
 
-    public Builder<T> setRequest(T request) {
-      this.request = request;
+    public Builder setRun(OperationRun run) {
+      this.run = run;
       return this;
     }
 
-    public Builder<T> setRun(OperationRun run) {
-      this.run = run;
+    public Builder setRunId(OperationRunId runId) {
+      this.runId = runId;
       return this;
     }
 
-    public Builder<T> setRunId(OperationRunId runId) {
-      this.runId = runId;
+
+    public Builder setPullAppsRequest(PullAppsRequest pullAppsRequest) {
+      this.pullAppsRequest = pullAppsRequest;
       return this;
     }
 
     /**
      * Validates input and returns a OperationRunDetail.
      */
-    public OperationRunDetail<T> build() {
+    public OperationRunDetail build() {
       if (runId == null) {
         throw new IllegalArgumentException("run id must be specified.");
       }
-      if (request == null) {
-        throw new IllegalArgumentException("Operation run request must be specified.");
-      }
       if (sourceId == null) {
         throw new IllegalArgumentException("Operation run source id must be specified.");
       }
       if (run == null) {
         throw new IllegalArgumentException("Operation run must be specified.");
       }
 
-      return new OperationRunDetail<>(runId, run, sourceId, principal, request);
+      if (!validateRequests()) {
```

Comment:
we should also be verifying somewhere that the OperationType matches the request, though it is kind of awkward now because they're in different classes. Perhaps we should change it back to have this extend OperationRun?

Reply:
Dont have any major concern now for changing it back as mainly generic types was making the builders awkward before. 
But I think checking the type from inner objects should be fine as it will only happen in one function. And the composition makes things a lot cleaner.
Anyway I will add a todo and take up the changes in a separate PR as I want to make a class which contains the mappings and provides utility methods. That would be useful for both run-details and runners.